### PR TITLE
sanitize behavior when estimator passed to apply() is a dataop

### PR DESCRIPTION
**Pull request by jeromedockes**
this PR fixes 2 issues that could happen when the estimator passed to `.skb.apply()` is a data_op: when checking the existence of some methods and when accessing the scikit-learn tags. it is kind of an edge case but there are valid use cases, like switching to a dummy or faster estimator in preview mode or setting a name on the estimator to retrieve it more easily on an unfitted learner.

### CODE REVIEW

File: skrub/_data_ops/tests/test_data_ops.py [hunk 1]
```diff
@@ -639,3 +633,92 @@ def test_get_vars():
...
+        X_a, y_a = make_regression(random_state=0)
+    else:
+        X_a, y_a = make_classification(random_state=0)
+    X_df = pd.DataFrame(X_a).rename(columns=str)
+    if has_preview:
+        X, y = skrub.X(X_df), skrub.y(y_a)
+    else:
+        X, y = skrub.X(), skrub.y()
+    if needs_data:
+
```
Comment (rcap107) [inline skrub/_data_ops/tests/test_data_ops.py:658]: Could you explain this condition?

File: skrub/_data_ops/tests/test_data_ops.py [hunk 2]
```diff
@@ -639,3 +633,92 @@ def test_get_vars():
...
+            return {}
+
+        def set_params(self):
+            return self
+
+    learner = skrub.var("a").skb.apply(Twice()).skb.make_learner()
+    assert learner.fit_transform({"a": 1}) == 2
+    xy_learner = learner.__skrub_to_Xy_pipeline__({})
+    assert xy_learner._estimator_type == "transformer"
+    if hasattr(xy_learner, "__sklearn_tags__"):
```
Comment (rcap107) [inline skrub/_data_ops/tests/test_data_ops.py:722]: is `hasattr` here for compatibility? I don't see conditions that should change this

**Comment by rcap107** [comment]
>  it is kind of an edge case but there are valid use cases, like switching to a dummy or faster estimator in preview mode or setting a name on the estimator to retrieve it more easily on an unfitted learner.

I feel like there should be an example of those use cases in the user guide because otherwise I think almost nobody will ever figure out those use cases on their own

It could be added at a later time, but still it should be there

**Comment by jeromedockes** [comment]
Thanks @rcap107 , I tried to answer the questions directly in comments in the code. I agree about the need for advanced examples (but about this PR specifically, I think someone asked about something similar in the skrub discord)
